# **Projet 3 - Prédiction de la consommation énergétique des bâtiments de Seattle**

 * Prédire la consommation énergétique des bâtiments non résidentiels de Seattle à partir de leurs caractéristiques structurelles.
 * Réaliser une analyse exploratoire afin d'identifier les principales tendances et anomalies des données.
 * Comparer plusieurs modèles de régression supervisée pour sélectionner le plus performant.
 * Identifier les variables ayant le plus d'influence sur la consommation énergétique.*
 

## **Contexte** 

* La ville de Seattle souhaite atteindre la neutralité carbone d'ici 2050.
* Les relevés énergétiques des bâtiments sont coûteux à réaliser et ne sont pas disponibles pour l'ensemble du parc immobilier.
* L'objectif est d'utiliser les données collectées en 2016 afin d'estimer la consommation énergétique des bâtiments non encore mesurés.*

## **Objectfs Data** 
- **Comprendre le jeu de données :** analyser la structure, les types de variables, les valeurs manquantes et la qualité des données.
- **Nettoyer et préparer les données :** sélectionner les bâtiments non résidentiels, traiter les incohérences et préparer les variables pour la modélisation.
- **Réaliser une analyse exploratoire (EDA) :** identifier les distributions, les corrélations et les facteurs susceptibles d'influencer la consommation énergétique.
- **Construire et comparer plusieurs modèles de régression** afin de prédire la consommation énergétique des bâtiments.
- **Interpréter le modèle retenu :** analyser l'importance des variables explicatives et identifier les principaux facteurs influençant la consommation énergétique.

## **Partie 1** 

### **Étape 1- Réalisez une analyse exploratoire**

A réaliser : 
- Une analyse descriptive des données, y compris une explication du sens des colonnes gardées, des arguments derrière la suppression de lignes ou de colonnes, des statistiques descriptives et des visualisations pertinentes.

Qelques pistes d'analyse :

* Identifier les colonnes avec une majorité de valeurs manquantes ou constantes en utilisant la méthode value_counts() de Pandas
* Mettre en evidence les différences entre les immeubles mono et multi-usages
* Utiliser des pairplots et des boxplots pour faire ressortir les outliers ou des batiments avec des valeurs peu cohérentes d'un point de vue métier 

Pour vous inspirer, ou comprendre l'esprit recherché dans une analyse exploratoire, vous pouvez consulter ce notebook en ligne : https://www.kaggle.com/code/pmarcelino/comprehensive-data-exploration-with-python. Il ne s'agit pas d'un modèle à suivre à la lettre ni d'un template d'analyses attendues pour ce projet. 

#### **Importation des librairies**

In [29]:
#importation des librairies

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

print("Importation des librairies OK")

Importation des librairies OK


#### **Importation du Dataset**

In [30]:
df_building_consumption = pd.read_csv("../data/2016_Building_Energy_Benchmarking.csv")

#### **Analyse Exploratoire**

##### **Standardisation des colonnes**

> * Adaptation et standardisation des noms de colonnes conformément aux bonnes pratiques.

In [31]:
# visualisation des colonnes de la base de données
df_building_consumption.columns

Index(['OSEBuildingID', 'DataYear', 'BuildingType', 'PrimaryPropertyType',
       'PropertyName', 'Address', 'City', 'State', 'ZipCode',
       'TaxParcelIdentificationNumber', 'CouncilDistrictCode', 'Neighborhood',
       'Latitude', 'Longitude', 'YearBuilt', 'NumberofBuildings',
       'NumberofFloors', 'PropertyGFATotal', 'PropertyGFAParking',
       'PropertyGFABuilding(s)', 'ListOfAllPropertyUseTypes',
       'LargestPropertyUseType', 'LargestPropertyUseTypeGFA',
       'SecondLargestPropertyUseType', 'SecondLargestPropertyUseTypeGFA',
       'ThirdLargestPropertyUseType', 'ThirdLargestPropertyUseTypeGFA',
       'YearsENERGYSTARCertified', 'ENERGYSTARScore', 'SiteEUI(kBtu/sf)',
       'SiteEUIWN(kBtu/sf)', 'SourceEUI(kBtu/sf)', 'SourceEUIWN(kBtu/sf)',
       'SiteEnergyUse(kBtu)', 'SiteEnergyUseWN(kBtu)', 'SteamUse(kBtu)',
       'Electricity(kWh)', 'Electricity(kBtu)', 'NaturalGas(therms)',
       'NaturalGas(kBtu)', 'DefaultData', 'Comments', 'ComplianceStatus',
       'Outlier

In [32]:
#Adaptation et standardisation des noms de colonnes conformément aux bonnes pratiques

df_building_consumption.columns = (
    df_building_consumption.columns
    .str.lower()                  # minúscules
    .str.strip()                  # remotion des espaces en début et fin de chaîne
    .str.replace(" ", "_")        # espace -> _
    .str.replace("-", "_")        # hífen -> _

)

In [33]:
# visualisation des colonnes après adaptation et standardisation

df_building_consumption.columns

Index(['osebuildingid', 'datayear', 'buildingtype', 'primarypropertytype',
       'propertyname', 'address', 'city', 'state', 'zipcode',
       'taxparcelidentificationnumber', 'councildistrictcode', 'neighborhood',
       'latitude', 'longitude', 'yearbuilt', 'numberofbuildings',
       'numberoffloors', 'propertygfatotal', 'propertygfaparking',
       'propertygfabuilding(s)', 'listofallpropertyusetypes',
       'largestpropertyusetype', 'largestpropertyusetypegfa',
       'secondlargestpropertyusetype', 'secondlargestpropertyusetypegfa',
       'thirdlargestpropertyusetype', 'thirdlargestpropertyusetypegfa',
       'yearsenergystarcertified', 'energystarscore', 'siteeui(kbtu/sf)',
       'siteeuiwn(kbtu/sf)', 'sourceeui(kbtu/sf)', 'sourceeuiwn(kbtu/sf)',
       'siteenergyuse(kbtu)', 'siteenergyusewn(kbtu)', 'steamuse(kbtu)',
       'electricity(kwh)', 'electricity(kbtu)', 'naturalgas(therms)',
       'naturalgas(kbtu)', 'defaultdata', 'comments', 'compliancestatus',
       'outlier

##### **Dimensions**

In [37]:
print("La base de données contient", df_building_consumption.shape[0], "lignes et", df_building_consumption.shape[1], "colonnes.")

La base de données contient 3376 lignes et 46 colonnes.


In [35]:
# On regarde comment un batiment est défini dans ce jeu de données 

df_building_consumption.head()

,osebuildingid,datayear,buildingtype,primarypropertytype,propertyname,address,city,state,zipcode,taxparcelidentificationnumber,...,electricity(kwh),electricity(kbtu),naturalgas(therms),naturalgas(kbtu),defaultdata,comments,compliancestatus,outlier,totalghgemissions,ghgemissionsintensity
0,1,2016,NonResidential,Hotel,Mayflower park hotel,405 Olive way,Seattle,WA,98101.0,0659000030,...,1.156514e+06,3946027.0,12764.52930,1276453.0,False,NaN,Compliant,NaN,249.98,2.83
1,2,2016,NonResidential,Hotel,Paramount Hotel,724 Pine street,Seattle,WA,98101.0,0659000220,...,9.504252e+05,3242851.0,51450.81641,5145082.0,False,NaN,Compliant,NaN,295.86,2.86
2,3,2016,NonResidential,Hotel,5673-The Westin Seattle,1900 5th Avenue,Seattle,WA,98101.0,0659000475,...,1.451544e+07,49526664.0,14938.00000,1493800.0,False,NaN,Compliant,NaN,2089.28,2.19
3,5,2016,NonResidential,Hotel,HOTEL MAX,620 STEWART ST,Seattle,WA,98101.0,0659000640,...,8.115253e+05,2768924.0,18112.13086,1811213.0,False,NaN,Compliant,NaN,286.43,4.67
4,8,2016,NonResidential,Hotel,WARWICK SEATTLE HOTEL (ID8),401 LENORA ST,Seattle,WA,98121.0,0659000970,...,1.573449e+06,5368607.0,88039.98438,8803998.0,False,NaN,Compliant,NaN,505.01,2.88


In [36]:
# Aperçu derniéres 5 lignes

df_building_consumption.tail()

,osebuildingid,datayear,buildingtype,primarypropertytype,propertyname,address,city,state,zipcode,taxparcelidentificationnumber,...,electricity(kwh),electricity(kbtu),naturalgas(therms),naturalgas(kbtu),defaultdata,comments,compliancestatus,outlier,totalghgemissions,ghgemissionsintensity
3371,50222,2016,Nonresidential COS,Office,Horticulture building,1600 S Dakota St,Seattle,WA,NaN,1624049080,...,153655.0000,5.242709e+05,3254.750244,3.254750e+05,True,NaN,Error - Correct Default Data,NaN,20.94,1.70
3372,50223,2016,Nonresidential COS,Other,International district/Chinatown CC,719 8th Ave S,Seattle,WA,NaN,3558300000,...,116221.0000,3.965461e+05,5537.299805,5.537300e+05,False,NaN,Compliant,NaN,32.17,2.01
3373,50224,2016,Nonresidential COS,Other,Queen Anne Pool,1920 1st Ave W,Seattle,WA,NaN,1794501150,...,525251.6875,1.792159e+06,39737.390630,3.973739e+06,False,NaN,Compliant,NaN,223.54,16.99
3374,50225,2016,Nonresidential COS,Mixed Use Property,South Park Community Center,8319 8th Ave S,Seattle,WA,NaN,7883603155,...,102248.0000,3.488702e+05,3706.010010,3.706010e+05,False,NaN,Compliant,NaN,22.11,1.57
3375,50226,2016,Nonresidential COS,Mixed Use Property,Van Asselt Community Center,2820 S Myrtle St,Seattle,WA,NaN,7857002030,...,126774.3984,4.325542e+05,7203.419922,7.203420e+05,False,NaN,Compliant,NaN,41.27,2.26


> * Les cinq premiers et 5 derniéres enregistrements correspondent à des hôtels situés à Seattle (WA) en 2016. 
> * Ils illustrent les principales variables du jeu de données : localisation, caractéristiques du bâtiment, consommation énergétique et émissions de gaz à effet de serre.


##### **Types de données**

In [50]:
# Types de données
df_building_consumption.dtypes

osebuildingid                        int64
datayear                             int64
buildingtype                           str
primarypropertytype                    str
propertyname                           str
address                                str
city                                   str
state                                  str
zipcode                            float64
taxparcelidentificationnumber          str
councildistrictcode                  int64
neighborhood                           str
latitude                           float64
longitude                          float64
yearbuilt                            int64
numberofbuildings                  float64
numberoffloors                       int64
propertygfatotal                     int64
propertygfaparking                   int64
propertygfabuilding(s)               int64
listofallpropertyusetypes              str
largestpropertyusetype                 str
largestpropertyusetypegfa          float64
secondlarge

##### **Doublons**

In [63]:
# Verification des doublons dans le jeu de données

df_building_consumption.duplicated().sum()

np.int64(0)

> **Observation :** 

> * Il n'y a pas de doublons

In [52]:
df_building_consumption.shape

(3376, 46)

##### **Valeurs manquantes**

In [61]:
# identification des valeurs manquantes

for col in df_building_consumption.columns:
    print(f"\n--- {col} ---")
    print(df_building_consumption[col].value_counts(dropna=False).head())


--- osebuildingid ---
osebuildingid
1    1
2    1
3    1
5    1
8    1
Name: count, dtype: int64

--- datayear ---
datayear
2016    3376
Name: count, dtype: int64

--- buildingtype ---
buildingtype
NonResidential          1460
Multifamily LR (1-4)    1018
Multifamily MR (5-9)     580
Multifamily HR (10+)     110
SPS-District K-12         98
Name: count, dtype: int64

--- primarypropertytype ---
primarypropertytype
Low-Rise Multifamily           987
Mid-Rise Multifamily           564
Small- and Mid-Sized Office    293
Other                          256
Warehouse                      187
Name: count, dtype: int64

--- propertyname ---
propertyname
Northgate Plaza            3
Lakeview                   2
South Park                 2
Garden Court Apartments    2
Central Park               2
Name: count, dtype: int64

--- address ---
address
2600 SW Barton St              4
2203 Airport Way S             4
309 South Cloverdale Street    4
500 5TH AVE                    2
3613 4th Ave S   

In [62]:
# Calcul du pourcentage moyen de valeurs manquantes par colonne

df_building_consumption["comments"].value_counts(dropna=False)

comments
NaN    3376
Name: count, dtype: int64

In [60]:
#creation d'un tableau récapitulatif des colonnes de la base de données

summary = pd.DataFrame({
    "dtype": df_building_consumption.dtypes,
    "non_null": df_building_consumption.count(),
    "missing": df_building_consumption.isna().sum(),
    "missing_pct": (
        df_building_consumption.isna().mean() * 100
    ).round(2),
    "n_unique": df_building_consumption.nunique()
})

summary.sort_values("missing_pct", ascending=False)

,dtype,non_null,missing,missing_pct,n_unique
comments,float64,0,3376,100.00,0
outlier,str,32,3344,99.05,2
yearsenergystarcertified,object,119,3257,96.48,65
thirdlargestpropertyusetype,str,596,2780,82.35,44
thirdlargestpropertyusetypegfa,float64,596,2780,82.35,501
secondlargestpropertyusetype,str,1679,1697,50.27,50
secondlargestpropertyusetypegfa,float64,1679,1697,50.27,1352
energystarscore,float64,2533,843,24.97,100
largestpropertyusetypegfa,float64,3356,20,0.59,3122
largestpropertyusetype,str,3356,20,0.59,56


> **Observation :** 

> * Les colonnes `comments` (100 %), `outlier` (99,05 %) et `yearenergystarcertified` (96,48 %) sont presque entièrement vides et seront étudiées afin de déterminer leur intérêt pour la modélisation. 
> * Les variables `secondlargestpropertyusetype` et `thirdlargestpropertyusetype`, ainsi que leurs surfaces associées, présentent également un nombre élevé de valeurs manquantes. En revanche, la majorité des autres variables comporte moins de 1 % de valeurs manquantes.

In [67]:
constant_columns = ["city", "state", "datayear"]

for col in constant_columns:
    print(f"\n--- {col} ---")
    print(df_building_consumption[col].value_counts(dropna=False))


--- city ---
city
Seattle    3376
Name: count, dtype: int64

--- state ---
state
WA    3376
Name: count, dtype: int64

--- datayear ---
datayear
2016    3376
Name: count, dtype: int64


In [65]:
df_building_consumption["state"].value_counts(dropna=False)

state
WA    3376
Name: count, dtype: int64

In [66]:
df_building_consumption["datayear"].value_counts(dropna=False)

datayear
2016    3376
Name: count, dtype: int64

##### **Suppression des colonnes inutilisables**

##### **Statistiques Descriptives**

In [48]:
describe = df_building_consumption.describe().T

describe.style.format(
    lambda x: f"{x:,.2f}".replace(",", " ").replace(".", ",")
)

,count,mean,std,min,25%,50%,75%,max
osebuildingid,"3 376,00","21 208,99","12 223,76","1,00","19 990,75","23 112,00","25 994,25","50 226,00"
datayear,"3 376,00","2 016,00","0,00","2 016,00","2 016,00","2 016,00","2 016,00","2 016,00"
zipcode,"3 360,00","98 116,95","18,62","98 006,00","98 105,00","98 115,00","98 122,00","98 272,00"
councildistrictcode,"3 376,00","4,44","2,12","1,00","3,00","4,00","7,00","7,00"
latitude,"3 376,00","47,62","0,05","47,50","47,60","47,62","47,66","47,73"
longitude,"3 376,00","-122,33","0,03","-122,41","-122,35","-122,33","-122,32","-122,22"
yearbuilt,"3 376,00","1 968,57","33,09","1 900,00","1 948,00","1 975,00","1 997,00","2 015,00"
numberofbuildings,"3 368,00","1,11","2,11","0,00","1,00","1,00","1,00","111,00"
numberoffloors,"3 376,00","4,71","5,49","0,00","2,00","4,00","5,00","99,00"
propertygfatotal,"3 376,00","94 833,54","218 837,61","11 285,00","28 487,00","44 175,00","90 992,00","9 320 156,00"


> **Observation :** 
> * Les statistiques descriptives montrent que le jeu de données contient 3 376 bâtiments. Les variables `datayear`, `city` et `state` sont concernent une seule campagne de collecte réalisée en 2016 à Seattle. 
> * Les caractéristiques physiques des bâtiments présentent une forte variabilité, notamment la surface totale (`propertygfatotal`) et le nombre d'étages (`numberoffloors`). 
> * Les variables de consommation énergétique et d'émissions de gaz à effet de serre suggérent la présence de valeurs extrêmes qui devront être analysées.   

     

In [39]:
df_building_consumption.info()

<class 'pandas.DataFrame'>
RangeIndex: 3376 entries, 0 to 3375
Data columns (total 46 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   osebuildingid                    3376 non-null   int64  
 1   datayear                         3376 non-null   int64  
 2   buildingtype                     3376 non-null   str    
 3   primarypropertytype              3376 non-null   str    
 4   propertyname                     3376 non-null   str    
 5   address                          3376 non-null   str    
 6   city                             3376 non-null   str    
 7   state                            3376 non-null   str    
 8   zipcode                          3360 non-null   float64
 9   taxparcelidentificationnumber    3376 non-null   str    
 10  councildistrictcode              3376 non-null   int64  
 11  neighborhood                     3376 non-null   str    
 12  latitude                       

> **Observation :** 
>  * Le jeu de données contient **3 376 observations** et **46 variables**. 
>  * L'aperçu fourni par `info()` met en évidence la présence de valeurs manquantes dans certaines variables, notamment `zipcode`, `numberofbuildings`, `totalghgemissions` et `ghgemissionsintensity`. 
>  * Les types de données sont cohérents avec la nature des variables et ne nécessitent pas de conversion particulière à ce stade.

     

# Modélisation 

### Import des modules 

In [40]:
#Selection
from sklearn.model_selection import (
    train_test_split,
    GridSearchCV, 
    cross_validate,
)
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error 
from sklearn.inspection import permutation_importance

#Preprocess
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, StandardScaler

#Modèles
from sklearn.dummy import DummyRegressor
from sklearn.linear_model import LinearRegression
from sklearn.svm import SVR
from sklearn.ensemble import RandomForestRegressor


### Feature Engineering

A réaliser : Enrichir le jeu de données actuel avec de nouvelles features issues de celles existantes. 

En règle générale : On utilise la méthode .apply() de Pandas pour créer une nouvelle colonne à partir d'une colonne existante. N'hésitez pas à regarder les exemples dans les chapitres de cours donnés en ressource

In [41]:
# CODE FEATURE ENGINEERING

### Préparation des features pour la modélisation

A réaliser :
* Si ce n'est pas déjà fait, supprimer toutes les colonnes peu pertinentes pour la modélisation.
* Tracer la distribution de la cible pour vous familiariser avec l'ordre de grandeur. En cas d'outliers, mettez en place une démarche pour les supprimer.
* Débarrassez-vous des features redondantes en utilisant une matrice de corrélation de Pearson. Pour cela, utiisez la méthode corr() de Pandas, couplé d'un graphique Heatmap de la librairie Seaborn 
* Réalisez différents graphiques pour comprendre le lien entre vos features et la target (boxplots, scatterplots, pairplot si votre nombre de features numériques n'est pas très élevé).
*  Séparez votre jeu de données en un Pandas DataFrame X (ensemble de feautures) et Pandas Series y (votre target).
* Si vous avez des features catégorielles, il faut les encoder pour que votre modèle fonctionne. Les deux méthodes d'encodage à connaitre sont le OneHotEncoder et le LabelEncoder

In [42]:
# CODE PREPARATION DES FEATURES

### Comparaison de différents modèles supervisés

A réaliser :
* Pour chaque algorithme que vous allez tester, vous devez :
    * Réaliser au préalable une séparation en jeu d'apprentissage et jeu de test via une validation croisée.
    * Si les features quantitatives que vous souhaitez utiliser ont des ordres de grandeur très différents les uns des autres, et que vous utilisez un algorithme de regression qui est sensible à cette différence, alors il faut réaliser un scaling (normalisation) de la donnée au préalable.
    * Entrainer le modèle sur le jeu de Train
    * Prédire la cible sur la donnée de test (nous appelons cette étape, l'inférence).
    * Calculer les métriques de performance R2, MAE et RMSE sur le jeu de train et de test.
    * Interpréter les résultats pour juger de la fiabilité de l'algorithme.
* Vous pouvez choisir par exemple de tester un modèle linéaire, un modèle à base d'arbres et un modèle de type SVM
* Déterminer le modèle le plus performant parmi ceux testés.

In [43]:
# CODE COMPARAISON DES MODELES

### Optimisation et interprétation du modèle

A réaliser :
* Reprennez le meilleur algorithme que vous avez sécurisé via l'étape précédente, et réalisez une GridSearch de petite taille sur au moins 3 hyperparamètres.
* Si le meilleur modèle fait partie de la famille des modèles à arbres (RandomForest, GradientBoosting) alors utilisez la fonctionnalité feature importance pour identifier les features les plus impactantes sur la performance du modèle. Sinon, utilisez la méthode Permutation Importance de sklearn.

In [44]:
# CODE OPTIMISATION ET INTERPRETATION DU MODELE